packages/jsii-python-runtime/src/jsii/runtime.py

```diff
@@ -0,0 +1,244 @@
+        # we're going to dynamically handle these.
+        class_properties, new_attrs = {}, {}
+        for key, value in attrs.items():
+            if isinstance(value, jsii_classproperty):
+                class_properties[key] = value
+            else:
+                new_attrs[key] = value
+        new_attrs["__jsii_class_properties__"] = class_properties
+
+        obj = type.__new__(cls, name, bases, new_attrs)
+
+        # We need to go through an implement the __set_name__ portion of our API.
+        for key, value in obj.__jsii_class_properties__.items():
+            value.__set_name__(obj, key)
+
+        _reference_map.register_type(obj.__jsii_type__, obj)
+
+        return obj
+
+    def __call__(cls, *args, **kwargs):
+        # Create our object at the JS level.
+        # TODO: Handle args/kwargs
+        # TODO: Handle Overrides
+        ref = cls.__jsii_kernel__.create(cls)
+
+        # Create our object at the Python level.
+        obj = cls.__class__.from_reference(cls, ref, *args, **kwargs)
+
+        # Whenever the object we're creating gets garbage collected, then we want to
+        # delete it from the JS runtime as well.
+        # TODO: Figure out if this is *really* true, what happens if something goes
+        #       out of scope at the Python level, but something is holding onto it
+        #       at the JS level? What mechanics are in place for this if any?
+        weakref.finalize(obj, cls.__jsii_kernel__.delete, obj.__jsii_ref__)
+
+        # Instatiate our object at the Python level.
+        if isinstance(obj, cls):
+            obj.__init__(*args, **kwargs)
+
+        return obj
+
+    def from_reference(cls, ref, *args, **kwargs):
+        obj = cls.__new__(cls, *args, **kwargs)
+        obj.__jsii_ref__ = ref
+
+        _reference_map.register_reference(obj.__jsii_ref__.ref, obj)
+
+        return obj
+
+    def __getattr__(obj, name):
+        if name in obj.__jsii_class_properties__:
+            return obj.__jsii_class_properties__[name].__get__(obj, None)
+
+        raise AttributeError(f"type object {obj.__name__!r} has no attribute {name!r}")
+
+    def __setattr__(obj, name, value):
+        if name in obj.__jsii_class_properties__:
+            return obj.__jsii_class_properties__[name].__set__(obj, value)
+
+        return super().__setattr__(name, value)
+
+
+# NOTE: We do this goofy thing so that typing works on our generated stub classes,
+#       because MyPy does not currently understand the decorator protocol very well.
+#       Something like https://github.com/python/peps/pull/242 should make this no
+#       longer required.
+if typing.TYPE_CHECKING:
+    jsii_property = property
+else:
+    # The naming is a little funky on this, since it's a class but named like a
+    # function. This is done to better mimic other decorators like @proeprty.
+    class jsii_property:
+
+        # We throw away the getter function here, because it doesn't really matter or
```

Comment:
In general, I wanted to have the generated code be as simple as possible, and to be able to be used to follow the structure of the library (but obviously not the implementation). What you're suggestion would basically be, instead of:

```python
class Foo(metaclass=JSIIMeta, jsii_type="something"):
    @jsii_property
    def my_property(self) -> str:
        ...
```

To do something like:

```python
class Foo(metaclass=JSIIMeta, jsii_type="something"):
    @property
    def my_property(self) -> str:
        # Most direct translation, this could be made slightly nicer
        # by import jsii.runtime.kernel and using that and removing
        # the __jsii_kernel__ attribute.
        return self.__jsii__kernel.get(self.__jsii_ref__, "my_property")
```

Generally I just think the first one looks better. It uses ``...`` to suggest that the implementation has been omitted (which it has, the implementation is off in TypeScript somewhere). It also means that bug fixes that don't affect the generated code can be made without requiring the entire JSII ecosystem to regenerate code, they can just pull in a new jsii-python-runtime (although that's probably unlikely, since the method call can be made to be roughly as simple as the property call itself).

Ultimately, it was primarily an aesthetics thing, and I vastly preferred the way generated code turned out when I was using decorators to implement the functionality. It lets the generated code sort of function more like a header file, and sort of matches (in spirit at least) the API of CFII where you embed the C header file in Python, and it creates an automagic Python object that has all of the C functions defined in that header on it. It's not a 1:1 mapping because the JSII is materially different, but it felt like the closest translation of that concept to me.

Reply:
In general, I wanted to have the generated code be as simple as possible, and to be able to be used to follow the structure of the library (but obviously not the implementation). What you're suggestion would basically be, instead of:

```python
class Foo(metaclass=JSIIMeta, jsii_type="something"):
    @jsii_property
    def my_property(self) -> str:
        ...
```

To do something like:

```python
class Foo(metaclass=JSIIMeta, jsii_type="something"):
    @property
    def my_property(self) -> str:
        # Most direct translation, this could be made slightly nicer
        # by import jsii.runtime.kernel and using that and removing
        # the __jsii_kernel__ attribute.
        return self.__jsii__kernel.get(self.__jsii_ref__, "my_property")
```

Generally I just think the first one looks better. It uses ``...`` to suggest that the implementation has been omitted (which it has, the implementation is off in TypeScript somewhere). It also means that bug fixes that don't affect the generated code can be made without requiring the entire JSII ecosystem to regenerate code, they can just pull in a new jsii-python-runtime (although that's probably unlikely, since the method call can be made to be roughly as simple as the property call itself).

Ultimately, it was primarily an aesthetics thing, and I vastly preferred the way generated code turned out when I was using decorators to implement the functionality. It lets the generated code sort of function more like a header file, and sort of matches (in spirit at least) the API of CFII where you embed the C header file in Python, and it creates an automagic Python object that has all of the C functions defined in that header on it. It's not a 1:1 mapping because the JSII is materially different, but it felt like the closest translation of that concept to me.